### PR TITLE
Refresh for compat with IJ 2020+

### PR DESCRIPTION
**Pull request by Tristan971**
Hi,

Am getting into Nix, and I do love my IJ so I thought it'd be cool to get this plugin refreshed (and later on improved upon).

So this is a first PR related to refreshing the project structure, which, thanks to Jetbrains making quite a few leaps in plugin development QoL, wasn't too difficult.

With this I'm able to `gradle build` the project which yields a usable plugin package at `build/distributions/NixIDEA-<version>.zip`. No idea how to wire automatic update to the marketplace however, so keep it manual as it was before for now.

Let me know your thoughts and apologies for the out-of-nowhere PR.

Cheers

Note: I'm no fan of Gradle personally, but alas it's the route Jetbrains has chosen for plugins, and it does allow almost all the Lexer/Parser setup and plugin packaging to be trivialized, so we might as well embrace the recommended way (from: https://github.com/JetBrains/intellij-platform-plugin-template )

### CODE REVIEW

**Comment by jonringer** [comment]
O.o this is a big update. Unfortunately, I'm not qualified to give a good PR review :(

**Comment by Tristan971** [comment]
That's fine, note that 99% of the changes is deleting files under `src/gen` which are... well... generated at build time and hence shouldn't have been here to begin with :slightly_smiling_face:

File: src/main/lang/Nix.bnf [hunk 1]
```diff
@@ -69,21 +69,21 @@
...
     /* search path, resolve first segment against NIX_PATH */
-    SPATH = "regexp:\<[a-zA-Z0-9\.\_\-\+]+(\/[a-zA-Z0-9\.\_\-\+]+)*\>"
+    SPATH = "regexp:<[a-zA-Z0-9._\-+]+(/[a-zA-Z0-9._\-+]+)*>"
 
     /* home path, resolve to relative users home */
-    HPATH = "regexp:\~(\/[a-zA-Z0-9\.\_\-\+]+)+"
+    HPATH = "regexp:~(/[a-zA-Z0-9._\-+]+)+"
     /* somewhere else */
-    URI = "regexp:[a-zA-Z][a-zA-Z0-9\+\-\.]*\:[a-zA-Z0-9\%\/\?\:\@\&\=\+\$\,\-\_\.\!\~\*\']+"
+    URI = "regexp:[a-zA-Z][a-zA-Z0-9+\-.]*:[a-zA-Z0-9%/?:@&=+$,\-_.!~*']+"
```
Comment (Tristan971) [inline src/main/lang/Nix.bnf:86]: All changes here were simply intellij complaining about unnecessary escaping

**Comment by Mic92** [comment]
cc @mrVanDalo for testing.

**Comment by jul1u5** [comment]
Hi, this seems very cool! I wanted to try it out on NixOS, but the `:runIde` Gradle task ran into an error:
```
> Task :runIde FAILED
 A problem occurred starting process 'command '/home/julius/.gradle/caches/modules-2/files-2.1/com.jetbrains/jbre/jbr-11_0_6-linux-x64-b765.25/jbr/bin/java''
```

The issue seems to be quite triviail: the Gradle task downloaded a JRE that doesn't work on NixOS because it isn't patched, e.g., the interpreter is set to `/lib64/ld-linux-x86-64.so.2` which isn't there in NixOS.

I came up with a very brittle solution that seems to work (IDEA launches successfuly): 
```bash
# Remove the downloaded JRE directory
$ rm -rf /home/julius/.gradle/caches/modules-2/files-2.1/com.jetbrains/jbre/jbr-11_0_6-linux-x64-b765.25

# Symlink the same directory to a patched JetBrains JDK from nixpkgs
$ ln -s ln -s $(nix-build '<nixpkgs>' -A jetbrains.jdk)/lib/openjdk /home/julius/.gradle/caches/modules-2/files-2.1/com.jetbrains/jbre/jbr-11_0_6-linux-x64-b765.25

# Run the previously failing task
$ ./gradlew runIde
```

Perhaps this could be improved so that people running NixOS would have better experience contributing to the project. Maybe it's possible to bypass downloading the JRE by specifying a local path in `build.gradle.kts`?

**Comment by Mic92** [comment]
@jul1u5 were you able to test the plugin?

**Comment by jul1u5** [comment]
@Mic92 Hi, the syntax highlighting and error reporting seemed to work fine. However, I didn't know how to test the linting, profile management and template features that are mentioned in the [README](https://github.com/NixOS/nix-idea/blob/ea2c5c8dc18a07ab98291a9e8518de95950ed764/README.md).

I was thinking, that maybe the development environment for NixOS could be implemented in another PR after this one is merged. For now we could just add the fhsuserenv file that you linked, or at least mention it in the README.

**Comment by Mic92** [comment]
If at least syntax highlight works, we should also publish a new version as I suspect in the current state no one can use it in idea with newer versions.

**Comment by Tristan971** [comment]
Ah yeah, I didn't try the "runIde" task, only focused on the "build" one, since that generates the zip-packaged plugin.

It's only a matter of adding a GitHub action for example replicating travis and artifacting it as a release I suppose.

Can do it, or let it for others to figure out ; I'm not familiar how publication to the Jetbrains marketplace happens however, and I'm not sure the project is set up to allow individuals to try that out either?

**Comment by Mic92** [comment]
Maybe @edwtjo has more insights into that.

**Comment by JojOatXGME** [comment]
@Tristan971 Are you working on a release job? Otherwiese, I could also take a look. My experiance with GitHub Actions is limited but uploading an artifact when certain tags are pushed sounds easy enough.

**Comment by Tristan971** [comment]
> @Tristan971 Are you working on a release job? Otherwiese, I could also take a look. My experiance with GitHub Actions is limited but uploading an artifact when certain tags are pushed sounds easy enough.

Hi @JojOatXGME, unfortunately I don't quite have the time to work on it right now, so by all means go ahead with that :bow:

**Comment by JojOatXGME** [comment]
I created pull request #16 about a release job.

**Pull request merged**